conbench/entities/hardware.py

```diff
@@ -114,10 +114,12 @@ class Cluster(Hardware):
 
     @classmethod
     def upsert(cls, **kwargs):
-        cluster = cls.first(name=kwargs["name"], info=kwargs["info"])
-        if cluster:
-            cluster.update(dict(optional_info=kwargs["optional_info"]))
-        else:
+        cluster = cls.first(
+            name=kwargs["name"],
+            info=kwargs["info"],
+            optional_info=kwargs["optional_info"],
+        )
+        if not cluster:
             cluster = cls.create(kwargs)
```

Comment:
There is a tiny chance that this create here can fail. Can you maybe try using https://github.com/conbench/conbench/blob/1f8a7f3a657c6028cb546bdc368af11d1afcb272/conbench/entities/_entity.py#L181?

Reply:
Oh wait, I just checked your method's code and it should work even without a unique constraint :) it will just fail to handle any race conditions, resulting in duplicate rows, just like my code.

Reply:
Okay, I switched to using `get_or_create()`.

The one thing it was missing: it went from `obj = cls(**props)` straight to `current_session.add(obj)`. But the Hardware objects generate the hash (checksum) column _before_ adding the object to the session, since that column is not nullable. So I added an overridable `post_init_pre_add()` method that will allow for custom initialization. It's not super pretty but I think it should handle any future use case.

Let me know what you think; I'm going to 🚢 this for now and we can revisit later.

Reply:
> I just checked your method's code and it should work even without a unique constraint :

:) yeah, cool!

> I added [...] for custom initialization

Ah! Could that also be in the `__init__()` method of `cls` so that `obj = cls(**props)` would run that logic? :shrug: :)

Reply:
>Could that also be in the __init__() method

Haha, this is what I get for trying to go too fast on a Friday. Yes, that would be a nice simplification.